js/packages/openinference-instrumentation-llama-index/src/patch.ts

```diff
@@ -0,0 +1,248 @@
+  OpenInferenceSpanKind,
+  SemanticConventions,
+} from "@arizeai/openinference-semantic-conventions";
+import {
+  RetrieverQueryEngineQueryMethodType,
+  RetrieverRetrieveMethodType,
+  QueryEmbeddingMethodType,
+  LLMChatMethodType,
+  LLMCompleteMethodType,
+} from "./types";
+import {
+  getExecContext,
+  handleError,
+  safelyGetDocumentAttributes,
+  safelyGetEmbeddingAttributes,
+  safelyGetLLMChatAttributes,
+  safelyGetLLMCompleteAttributes,
+} from "./utils";
+
+function setSpanAttributes(params: {
+  span: Span;
+  attributes: Attributes | null;
+}) {
+  if (params.attributes != null) {
+    params.span.setAttributes(params.attributes);
+  }
+}
+
+export function patchQueryEngineQueryMethod(
+  original: RetrieverQueryEngineQueryMethodType,
+  tracer: Tracer,
+) {
+  return function (
+    this: unknown,
+    ...args: Parameters<RetrieverQueryEngineQueryMethodType>
+  ) {
+    const span = tracer.startSpan(`query`, {
+      kind: SpanKind.INTERNAL,
+      attributes: {
+        [SemanticConventions.OPENINFERENCE_SPAN_KIND]:
+          OpenInferenceSpanKind.CHAIN,
+        [SemanticConventions.INPUT_VALUE]: args[0].query,
+        [SemanticConventions.INPUT_MIME_TYPE]: MimeType.TEXT,
+      },
+    });
+
+    const execContext = getExecContext(span);
+
+    const execPromise = safeExecuteInTheMiddle<
+      ReturnType<RetrieverQueryEngineQueryMethodType>
+    >(
+      () => {
+        return context.with(execContext, () => {
+          return original.apply(this, args);
+        });
+      },
+      (error) => handleError(span, error),
+    );
+
+    const wrappedPromise = execPromise.then((result) => {
+      span.setAttributes({
+        [SemanticConventions.OUTPUT_VALUE]: result.response,
+        [SemanticConventions.OUTPUT_MIME_TYPE]: MimeType.TEXT,
+      });
+      span.end();
+      return result;
+    });
+    return context.bind(execContext, wrappedPromise);
+  };
+}
+
+export function patchRetrieveMethod(
+  original: RetrieverRetrieveMethodType,
+  tracer: Tracer,
+) {
+  return function (
+    this: unknown,
+    ...args: Parameters<RetrieverRetrieveMethodType>
+  ) {
+    const span = tracer.startSpan(`retrieve`, {
+      kind: SpanKind.INTERNAL,
+      attributes: {
+        [SemanticConventions.OPENINFERENCE_SPAN_KIND]:
+          OpenInferenceSpanKind.RETRIEVER,
+        [SemanticConventions.INPUT_VALUE]: args[0].query,
+        [SemanticConventions.INPUT_MIME_TYPE]: MimeType.TEXT,
+      },
+    });
+
+    const execContext = getExecContext(span);
+
+    const execPromise = safeExecuteInTheMiddle<
+      ReturnType<RetrieverRetrieveMethodType>
+    >(
+      () => {
+        return context.with(execContext, () => {
+          return original.apply(this, args);
+        });
+      },
+      (error) => handleError(span, error),
+    );
+
+    const wrappedPromise = execPromise.then((result) => {
+      setSpanAttributes({
+        span: span,
+        attributes: safelyGetDocumentAttributes(result),
+      });
```

Comment:
New way to set span attributes and check if safely... method returns null. Is this a good way to go about this?